### ISSUE TITLE
Add "ctx.forward" for forwarding (bypassing) requests

### ISSUE THREAD

**Issue by kettanaito** [mswjs/msw#676]
## What

A new `ctx.forward` utility may be added to forward the currently captured request to its original destination (bypass).

## Why

- See #655 

Currently bypassing is done via implicit/explicit return in the resolver. This is not intuitive and deterministic action, which also makes it hard to distinguish when the user _forgot_ to return a mocked response, and when they _meant_ to bypass intentionally. 

With the introduction of this feature, the library could reliably print a warning when you have a handler, but return no response from the resolver—it becomes safe to assume that you forgot to do that. 

## API

```js
rest.get('/user', (req, res, ctx) => {
  if (anything) {
    return ctx.forward()
  }

  return res(ctx.text('mocked'))
})
```

**Comment by kettanaito** [comment]
I wonder if this can be substituted at the moment by using `ctx.fetch`?

```js
rest.get('/user', (req, res, ctx) => {
  if (anything) {
    return ctx.fetch(req)
  }

  return res(ctx.text('mocked'))
})
```

I suppose not, as the return signature of `ctx.fetch` isn't of the `MockedResponse` type.

**Comment by marcosvega91** [comment]
If I a handler for a request is found with no response MSW will not bypass the request anymore except if the user will use `forward` utility, right ?

**Comment by kettanaito** [comment]
@marcosvega91, correct. The `ctx.forward` utility becomes the only way to bypass a request in a handler. Returning no mocked response will raise a warning because that means the user forgot to handle that request at all. There will be only two possibilities to _handle_ a request:

- Return a mocked response by `return res()`.
- Forward (bypass) a request by `ctx.forward()`.

Anything else is treated as an unhandled request.

**Comment by marcosvega91** [comment]
it will be the default behaviour ?. For example I could use some library that I don't want to mock because I don't know well the ecosystem ( using firebase as an example). Is there will be a way to auto-forward requests ?

**Comment by kettanaito** [comment]
MSW behavior won't change concerning requests that have no handlers. Those requests will be forwarded just as they are now. The change comes when you define a handler for a request. To forward such _handled_ request, you would have to explicitly return `ctx.forward`. I think this lies within the current behavior of default forwarding, just changes the public API and makes the forwarding more explicit.

**Comment by adarnon** [comment]
I see a great use case for this together with the default `onUnhandledRequest: 'bypass'` setting option. Currently there's no way to mark certain requests as "should be bypassed", for tests where you want the real backend to be called and don't want MSW to touch anything. What about doing this at the server level - `server.bypass(...matcher)`?

**Comment by kettanaito** [comment]
Hey, @adarnon.

There _is_ a way to bypass any request in the handler directly right now:

```js
rest.get('/user', (req, res, ctx) => {
  // Skip this handler and perform the request as-is
  // if the request URL contains the "skip" query parameter.
  if (req.url.searchParams.get('skip')) {
    return
  }

  return res(ctx.json({ id: 1 }))
})
```

> Read more in [Conditional response](https://mswjs.io/docs/basics/response-resolver#conditional-response).

Both explicit and implicit `return` in the resolver that doesn't return a mocked response composition (`res`) is treated as the instruction to perform that captured request as-is. Doing so will not produce any `onUnhandledRequest` warnings, as the request was handled, just not resolved to a mocked response.

What this task is about is to introduce an _explicit_ API to state that you wish to perform the request as-is, opposed to having to rely on how the library interprets the `return` statement. 

```js
if (req.url.searchParams.get('skip')) {
  return res(ctx.forward())
}
```

**Comment by Oluwasetemi** [comment]
@kettanaito I want to jump on this. What do I need to do?

**Comment by Oluwasetemi** [comment]
how will the forwarding work? @kettanaito

**Comment by Oluwasetemi** [comment]
```js
export const forward = () => {
  return (res) => {
    return res
  }
}
```

Will it just return the request to its destination?

**Comment by IanVS** [comment]
>Doing so will not produce any onUnhandledRequest warnings, as the request was handled, just not resolved to a mocked response.

It might not be an `onUnhandledRequest` warning, but I do get this in my console:

```
[MSW] Expected a mocking resolver function to return a mocked response Object, but got: undefined. Original response is going to be used instead.
```

I'd love to be able to mark paths as explicitly bypassed / passthrough.  I'd suggest though, instead of calling it `ctx.forward()`, perhaps to use the same word as [mirage](https://miragejs.com/api/classes/server/#passthrough), and call it `ctx.passthrough()`, since at least to me, that makes the behavior more clear.  To my mind, `forward` makes it seem like sending it somewhere that it wasn't sent initially, like forwarding an email to a new recipient.

**Comment by Xunnamius** [comment]
Hi! Any updates on this? Trying out MSW in my project and so far so good! Except my terminal is polluted with a bajillion warnings about the responses I'm bypassing _on purpose_ (i.e. returning `undefined` as described in [the documentation](https://mswjs.io/docs/basics/response-resolver#conditional-response)). If it matters, I'm also using `onUnhandledRequest: 'error'` to ensure no requests slip by. To preserve my sanity, I've monkey-patched `console.warn` to ignore calls with strings that start with `"[MSW]"`, but that's a temp fix at best.

I also spotted what seems like a duplicate issue to this one: https://github.com/mswjs/msw/issues/901

**Comment by kettanaito** [comment]
@Oluwasetemi,

> how will the forwarding work?

Currently, a request is bypassed for the following reasons:

1. It has the `x-msw-bypass` request header set;
2. It does not have any associated request handler;
3. It has a handler but the handler returned no mocked response.

What we're trying to solve with `ctx.forward` is to give us a context to differentiate between 3 (handler without response) and an intentional bypassing of requests. 

The mocked response is retrieved in `getResponse`, which opens us to these two potential branches of logic:

https://github.com/mswjs/msw/blob/6b19df4e5bd1e8614eb12fa6edecc41514769591/src/utils/handleRequest.ts#L75-L103

I'd leave the `!handler` processing as-is, it's correct: when there's no handler the `onUnhandledRequest` logic executes without any hard-coded warnings on our side. It's the `!response` part that we need to adjust to differentiate between unintentional/intentional lack of mocked response. 

I think the least intrusive implementation here would be for the `ctx.forward` to return a symbol denoting the intention to bypass the request:

```js
export const RESPONSE_FORWARD = new Symbol('RESPONSE_FORWARD')

export function forward(res) {
  return (res) => {
    return RESPONSE_FORWARD
  }
}
```
> I consider using a symbol so that we treat it as an internal implementation detail entirely, forbidding the end-user to just return a fixed string for the magic to happen.

Then in `handleRequest` we can check for that symbol and print the warning only if it's not the case:

```js
import { RESPONSE_FORWARD } from '../../ctx/forward'

const { response } = await getResponse(...)

if (!response || response === RESPONSE_FORWARD) {
  if (!response) {
    devUtils.warn('Warn about unintentional empty handler')
  }

  // otherwise, don't.
}
```

We may iterate on this afterward, introducing a more consistent model to communicate custom response behaviors to the request handling pipeline.

**Comment by kettanaito** [comment]
@IanVS,

>  I'd suggest though, instead of calling it ctx.forward(), perhaps to use the same word as mirage, and call it ctx.passthrough(), since at least to me, that makes the behavior more clear.

Thanks for the suggestion! What do you think about `ctx.bypass`? We are using the "bypass" terminology everywhere so I'd favor this particular name. I don't have anything against `ctx.passthrough` if it's a more conventional name, though.

**Comment by IanVS** [comment]
I don't have a strong opinion between bypass and passthrough, either one sounds reasonable, and would be an improvement over `forward` IMHO.

**Comment by kettanaito** [comment]
Hey, @Xunnamius, there are quite a lot of tasks to tackle so some, inevitably, get overlooked. This one isn't the case but I wanted to share that managing a project of this scale with a few people on board is quite time-demanding, so I hope for your understanding.

> Alternatively, your company can consider sponsoring MSW to become our [first Golden sponsor](https://github.com/sponsors/mswjs/sponsorships?sponsor=kettanaito&tier_id=121738&preview=false). With this tier we will prioritize features and issues that'd make your workflow first-grade. 

That being said, I've outlined the initial implementation proposal of `ctx.forward` in the comment above. I encourage you to take the initiative and help us build it by opening a pull request (make sure to align with @Oluwasetemi beforehand). Otherwise, this feature will be worked on whenever people have time. 

> I also spotted what seems like a duplicate issue to this one: #901

That's a great observation! I suppose the only difference there is when you wish to transform the result of `ctx.fetch` and then return it:

```js
const response = await ctx.fetch(req)
response.headers.set('x-custom', 'yes')
return response
```

The culprit here is that request handlers expect a `MockedResponse` abstraction, which is not compatible with the `Response` instance returned by `fetch`.

**Comment by Xunnamius** [comment]
@kettanaito Thank you for your kind response. I didn't mean to imply any sense of urgency, there's no rush, and I'm grateful for your work on this. I'm already using MSW successfully in a few projects and it's saved me loads of time. I'll look into submitting a PR in accordance with your instructions above  🙂

**Comment by rtalvarez** [comment]
Hey @Oluwasetemi, are you still working on this?

I'd like to help tackle this, sounds interesting. Thanks for the context @kettanaito

**Comment by Oluwasetemi** [comment]
Yes @rtalvarez .

**Comment by Oluwasetemi** [comment]
@rtalvarez no I am not. You can work on it.

**Comment by kettanaito** [comment]
## Released: v0.40.0 🎉

This has been released in v0.40.0!

- 📄 [**Release notes**](https://github.com/mswjs/msw/releases/tag/v0.40.0)
- 📦 [npm package](https://www.npmjs.com/package/msw/v/0.40.0)

Make sure to always update to the latest version (`npm i msw@latest`) to get the newest features and bug fixes.

---

_Predictable release automation by [@ossjs/release](https://github.com/ossjs/release)_.
